File: MyAppOne/MyAppOne/VC/Initial/MAOInitialViewController.m
//
//  MAOInitialViewController.m
//  MyAppOne
//
//  Created by Julio Castillo on 9/1/19.
//  Copyright © 2019 iOS-accelerator. All rights reserved.
//

#import "MAOInitialViewController.h"
#import "MAOListViewController.h"
#import "../../Services/MARequestService.h"
#import "../ListCustom/MAOListCustomViewController.h"
#import "../../Model/MAOListViewControllerModel.h"

@interface MAOInitialViewController ()

@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;
@property (nonatomic, strong) NSMutableArray<MAOListViewControllerModel *> *arrayModels;
@property (weak, nonatomic) IBOutlet UIImageView *resultIcon;

typedef enum {
    TRACK_ID,
    RELEASE_DATE
} OrderType;

@end

@implementation MAOInitialViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.activityIndicator setHidden:TRUE];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.navigationController setNavigationBarHidden:YES animated:YES];
}

- (IBAction)onClickSelection:(id)sender {
    [self.activityIndicator startAnimating];
    [self.activityIndicator setHidden:FALSE];
    
    [MARequestService fetchDataFromUrl:[NSURL URLWithString:@"https://itunes.apple.com/search?term=jack+johnson"]
           success:^(NSArray *dataArray) {
               NSLog(@"Data: %@",dataArray);
               dispatch_async(dispatch_get_main_queue(), ^{
                   [self.resultIcon setImage:[UIImage imageNamed:@"okIcon"]];
                   [self.activityIndicator stopAnimating];
                   [self.activityIndicator setHidden:TRUE];
               });
               dataArray = [dataArray valueForKey:@"results"];
               self.arrayModels = [[NSMutableArray<MAOListViewControllerModel *> alloc] init];
               for (id itemArray in dataArray) {
                   MAOListViewControllerModel *item = [[MAOListViewControllerModel alloc] initFromDictionary:itemArray];
                   [self.arrayModels addObject:item];
               }
           } error:^(NSError *error) {
               dispatch_async(dispatch_get_main_queue(), ^{
                   [self.resultIcon setImage:[UIImage imageNamed:@"errorIcon"]];
                   [self.activityIndicator stopAnimating];
                   [self.activityIndicator setHidden:TRUE];
               });
               [self showDialogWithTitle:@"Atención" withMessage:[error localizedDescription]];
           }];
}

- (IBAction)btnShowByReleaseTrackAsc:(UIButton *)sender {
    [self goToListShowDesc:FALSE orderByType:RELEASE_DATE];
}

- (IBAction)btnShowByReleaseTrackDesc:(UIButton *)sender {
    [self goToListShowDesc:TRUE orderByType:RELEASE_DATE];
}

- (IBAction)btnShowByTrackIdAsc:(UIButton *)sender {
    [self goToListShowDesc:FALSE orderByType:TRACK_ID];
}

- (IBAction)btnOrderByTrackIdDesc:(UIButton *)sender {
    [self goToListShowDesc:TRUE orderByType:TRACK_ID];
}

-(void) goToListShowDesc:(BOOL) showDesc orderByType:(int) type {
    if(_arrayModels != nil || [_arrayModels count] != 0) {
        NSArray *sortedArray = [_arrayModels sortedArrayUsingComparator: ^(MAOListViewControllerModel* obj1, MAOListViewControllerModel* obj2) {
            
            if(type == TRACK_ID) {
                NSNumber *obj1TrackId = [obj1 trackId];
                NSNumber *obj2TrackId = [obj2 trackId];
                
                if(showDesc) {
                    return [obj2TrackId compare:obj1TrackId];
                } else {
                    return [obj1TrackId compare:obj2TrackId];
                }
            } else if(type == RELEASE_DATE) {
                NSDate *obj1ReleaseDate = [obj1 releaseDate];
                NSDate *obj2ReleaseDate = [obj2 releaseDate];
                
                if(showDesc) {
                    return [obj2ReleaseDate compare:obj1ReleaseDate];
                } else {
                    return [obj1ReleaseDate compare:obj2ReleaseDate];
                }
            }
            return NSOrderedSame;
        }];
        
        MAOListCustomViewController *listView = [[MAOListCustomViewController alloc] initWithModel:sortedArray];
        [self.navigationController pushViewController:listView animated:YES];
    } else {
        [self.resultIcon setImage:[UIImage imageNamed:@"errorIcon"]];
        [self showDialogWithTitle:@"Atención" withMessage:@"Primero debes cargar los datos"];
    }
}

-(void) showDialogWithTitle:(NSString *)title withMessage:(NSString *)message {
    UIAlertController* alert = [UIAlertController alertControllerWithTitle:title
                                                                   message:message
                                                            preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction* defaultAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault
                                                          handler:^(UIAlertAction * action) {}];
    
    [alert addAction:defaultAction];
    [self presentViewController:alert animated:YES completion:nil];
}

@end

File: MyAppOneSwift/MyAppOneSwift/VC/Initial/MAOInitialViewController.swift
//
//  MAODetailViewController.swift
//  MyAppOneSwift
//
//  Created by Roman Guarino on 25/01/2019.
//  Copyright © 2019 Román Guarino. All rights reserved.
//

import UIKit
import RGNetworking
import ProgressHUD

class MAOInitialViewController: UIViewController {

    @IBOutlet weak var resultIcon: UIImageView!
    var arrayModels: Array<MAOListViewControllerModel> = []
    
    enum OrderType {
        case TRACK_ID
        case RELEASE_DATE
    }
    
    struct Result : Decodable {
        let resultCount: Int
        let results: [MAOListViewControllerModel]
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.navigationController?.setNavigationBarHidden(true, animated: animated)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        self.navigationController?.setNavigationBarHidden(false, animated: animated)
    }

    @IBAction func onClickSelection(_ sender: UIButton) {
        ProgressHUD.show()
        
        RGNetworking.fetchData(from: URL.init(string: "https://itunes.apple.com/search?term=jack+johnson"), success: { (data) in
            DispatchQueue.main.async {
                if let data = try? JSONDecoder().decode(Result.self, from: data!) {
                    ProgressHUD.showSuccess()
                    self.resultIcon.image = UIImage.init(named: "okIcon")
                    self.arrayModels = data.results
                } else {
                    ProgressHUD.showError("Se produjo un error al procesar los datos")
                    self.resultIcon.image = UIImage.init(named: "errorIcon")
                }
            }
        }) {(error) in
            DispatchQueue.main.async {
                self.resultIcon.image = UIImage.init(named: "errorIcon")
                print("Error al obtener los datos: \(error!.localizedDescription)")
                ProgressHUD.showError("Se produjo un error al obtener los datos")
            }
        }
    }
    
    @IBAction func btnShowByReleaseTrackAsc(_ sender: UIButton) {
        self.goToList(showDesc: false, orderByType: OrderType.RELEASE_DATE)
    }
    
    @IBAction func btnShowByReleaseTrackDesc(_ sender: UIButton) {
        self.goToList(showDesc: true, orderByType: OrderType.RELEASE_DATE)
    }
    
    @IBAction func btnShowByTrackIdAsc(_ sender: UIButton) {
        self.goToList(showDesc: false, orderByType: OrderType.TRACK_ID)
       
    }
    
    @IBAction func btnShowByTrackIdDesc(_ sender: UIButton) {
        self.goToList(showDesc: true, orderByType: OrderType.TRACK_ID)
    }
    
    func goToList(showDesc: Bool, orderByType: OrderType) {
        if(arrayModels.count != 0) {
            let sortedArray = arrayModels.sorted { (obj1, obj2) -> Bool in
                if(OrderType.TRACK_ID == orderByType) {
                    if(showDesc) {
                        return obj1.trackId! > obj2.trackId!
                    } else {
                        return obj1.trackId! < obj2.trackId!
                    }
                } else {
                    if(showDesc) {
                        return obj1.releaseDate! > obj2.releaseDate!
                    } else {
                        return obj1.releaseDate! < obj2.releaseDate!
                    }
                }
            }
            
            let nav = UINavigationController(rootViewController: self)
            UIApplication.shared.keyWindow?.rootViewController = nav
            let vc = MAOListViewController.init(arrayModel: sortedArray)
            nav.pushViewController(vc, animated: true)
        } else {
            self.resultIcon.image = UIImage.init(named: "errorIcon")
            showDialog(title: "Atención", message: "Primero debes cargar los datos")
        }
    }
    
    func showDialog(title: String, message: String) {
        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let alertAction = UIAlertAction(title: "Ok", style: .cancel, handler: nil)
        alertController.addAction(alertAction)
        present(alertController, animated: true, completion: nil)
    }
}
